### ISSUE TITLE
Allow partial mixing of concrete values with matchers

### ISSUE THREAD

**Issue by eyalroth** [mockito/mockito#1398]
I'm pretty sure we're all familiar with the dreadful message "Invalid use of argument matchers" which is caused (most of the time) due to mixing of concrete values and matchers. After using Mockito for several years, I gotta say that this is one of the most annoying drawbacks of the framework.

Recently I've been getting more and more into the code of the framework, and started wondering why this restriction is there in the first place. I looked it up, and found [this thread](https://groups.google.com/forum/#!topic/mockito/4W-PFmG5jTM) on the Google mailing list, in which a guy named Brice (@bric3 ?) explains the reasoning behind this in a fairly simple manner.

In short, the matchers cannot be passed to the mock invocation, since they have a different type than that of the arguments. The way Mockito is handling this is by "registering" the matchers prior to the invocation on a local-thread stack, which the mock-handler later inspects upon the method invocation. When only matchers are used, it is easy to match each matcher to a parameter of the method--by simply matching the stack to the parameter list--but when concrete values are mixed in, this becomes ambiguous.

I would like to suggest a middle-ground solution that will probably cover (at least) 80% of the cases - How about assuming that each `null` argument a method is invoked with is supposed to correspond to a matcher, while the rest will be treated as concrete values? This way, only cases where `null` is passed "concretely" alongside matchers will be invalid. For example:
```java
when(myMock.foo(any(), null)) // invalid
when(myMock.foo(any(), "1")) // valid
```

There would also be a problem with primitive types, as matchers cannot be passed to the method as `null`s, but in this case they could be passed as default values and then an invocation will be considered invalid if the number of default primitive values passed to the method exceeds the number of matchers of said primitive type on the stack.

Thoughts?

**Comment by TimvdLippe** [comment]
Yes, this is indeed one of the gotchas I consider most negative of Mockito. That said, I think solving this with the current implementation is near impossible. Matchers currentlly return default values, to make sure method invocations compile, while in the background we register the matcher. There is technically no way we can differentiate between a regular value and the default value of a matcher.

In #1264 I lifted this restriction to the type-level instead: https://github.com/mockito/mockito/blob/c7a326c85d7eb446f29b2e772d9ec218ab0d91cb/subprojects/lambda/src/main/java/org/mockito/OngoingStubbingFunction.java#L18-L26

That approach does allow us to differentiate and make this requirement more explicit.

**Comment by ChristianSchwarz** [comment]
@eyalroth Your proposal to distingush matchers from plain values by default values is doable, but as you already said there is still a hole when plain-values have the same value as the matcher return value. This may become more problematic in case mockito decides to return non null values for machers to support e.g. Kotlin. 

I see 2 different solutions (both costly):
 * Provide a IDE or compiler plugin that replace the plain-values to there corresponding matcher representation: `when(max(1,eq(3)))` to` when(max(eq(1),eq(3)))`, I think this is out of the mockito scope.
 * In case a "Invalid use of argument matchers" is detected, mockito could analyse the bytecode around the `when` and determine which argument was feeded by an matcher and which not. This would probably slow down the test a lot.

**Comment by eyalroth** [comment]
@TimvdLippe I took a look over your PR and it looks like it ain't gonna become the go-to way of mocking in Mockito anytime soon. I'm guessing that at it current state it doesn't support Scala, which is what I'm using in my current project, so I'm biased :)

@ChristianSchwarz It's good that you've mentioned the case with Kotlin; this may actually be the solution here (for the most part).

Mockito is pretty good at **mocking** objects, which is exactly what the matchers need - imagine that instead of returning `null`, the matcher methods will return a new "empty" mock of the appropriate type, which will be registered in the stack alongside the matcher (the stack will now contain pairs of mocks and matchers).

The only problem would be again with primitives. In this case I think it is sufficient enough to return a "default" value - probably not the actual default of the primitive but an arbitrary instance of it; say, with `long` the arbitrary instance could be `9545892384298471L`. This way, there is a small chance for a collision between an actual value and a matcher-default value, unless one mixes matchers and actual values with `boolean` or `byte` primitives, which I believe is quite an esoteric case.

**Comment by ChristianSchwarz** [comment]
@eyalroth

Returning "marker-mocks" sounds like good idea, I guess the inline-mockmaker is a requirement here in order to mock every kind of type (enum, non static inner-classes, final classes, ... ). The inline-mock-maker should be the default then.

Primitive default values with a abitrary value can help to distinguesh but may leave some cases that won't work, e.g. for primitives with a small value range like byte or boolean. In this cases an analysis of the argument-types order may help to distiguish . 

Examples to distinguish matchers and values by only looking into the argument type order:
 * Distinction possible: `foo(int, String)` when called with `foo(eq(5),"Hello")`
 * Distinction impossible:  foo(int,int) when called with `foo(eq(5),8)`

**Comment by ChristianSchwarz** [comment]
@TimvdLippe 
@bric3
@szpak 
@mockitoguy 
@mockitocore

What do you guys think of a limited support for mixed values/matchers? 
There is a good chance that Mockito can support mixing for more than 90% of all method calls (just a pessimistic guess). The famous "Invalid use of argument matchers" can become a corner case, were users are still forced to apply matchers everywhere. This would make test code even better / more readable.

I see currently following expansion stages:

 * support for mixing primitive values and build in mockito primitve matchers ( `eq(..)`, `any###() `) using default matcher return values
 * support for mixing primitive/ object values and build in mockito matchers, default matcher return values / mocks
 * improved matcher resolution using additional method signature analysis

**Comment by TimvdLippe** [comment]
The core problem I see is that it is impossible to distinguish a primitive value returned by a matcher to a primitive value passed in from the user. Given this inability to detect, we can only rely on heuristics and such heuristics will get it wrong. I think this would actually worsen the situation, as we now move these errors to an even more complicated feature to debug.

That said, I think the only fully compatible issue to lift these restrictions to the type-level, which my PR does. Agreed that this PR is still a long way from merging, but we have had this problem for a while and I do not think shipping a half-baked solution in the mean time is going to improve the state of affairs.

Therefore, I am against incorporating a solution that does not work all the time and would instead want to devote time to solutions that are proven to be working.

**Comment by ChristianSchwarz** [comment]
The proposed solution allows to distiguish values from matchers in the most uses cases. Those cases that can't be distiguished will still throw a "Invalid use of argument matchers".  The user is then (like now) forced to use matchers only. From a user perspective it is getting better.

**Comment by TimvdLippe** [comment]
> From a user perspective it is getting better.

I respectfully disagree. Right now the problem is strict in all cases: we disallow all mixing. This is clearly documented [in our JavaDoc including examples](http://static.javadoc.io/org.mockito/mockito-core/2.18.3/org/mockito/Mockito.html#argument_matchers). Relaxing this requirement would be a lot more difficult to document clearly and will introduce confusion.

Yes, agreed fewer cases will trigger the warning, however right now we are at least strict in it. I prefer a strict solution over a solution that works in a majority of the time, but increase ambiguity in other cases.

**Comment by ChristianSchwarz** [comment]
Well I think doc/confusion problem can be solved, so even John Doe will know what to do 😎. 
I doesn't need to be over complicated:

> If you get a "Invalid use of argument matchers" use matchers everywhere!

**Comment by rdp** [comment]
I suggested something similar once.  So obviously +1 from me.  The question is would it cause more confusion than not...but it certainly might clear up code, especially if you have multiple parameters and have `eq` in front of...almost...all...of...them.  https://groups.google.com/forum/#!topic/mockito/4W-PFmG5jTM :)

**Comment by mockitoguy** [comment]
If we can improve the user experience at some cost of internal complexity we should do it.

1. Are we improving the user experience?
2. What's the cost/complexity?

1) I think we are. If the exception becomes less likely to occur it improves user experience. We can document that "Mockito tries hard to guess the types but...". This is a pattern that we already leverage elsewhere - InjectMocks, MockitoHamcrest.argThat, etc. I agree with @TimvdLippe that we should not regress the existing UX. If we are detecting types automatically in 50% of cases but we are guessing them wrong in the other 50%, triggering confusing behavior of framework, that is *not* an "improvement" of UX.
2) Can someone try to hack it so that we can see? Please include very clear Test class that demonstrates improved behavior and the cases that we still don't support.

Mockito passed the point where we can easily make bulletproof changes/improvements. We squeezed as much as we could from the Java type system ;) It is OK to introduce partial solutions to use cases, given acceptable complexity + improved UX. We have been doing that for years.

**Comment by ChristianSchwarz** [comment]
@mockitoguy her you can find a hack for type `int` only, same can be applied to other types too: #1441

**Comment by eyalroth** [comment]
@mockitoguy @ChristianSchwarz  I'm trying to work on this but the tests do not pass on my machine: https://groups.google.com/forum/#!topic/mockito-dev/tf21JAiG0w8

Care to help?

**Comment by oleksiyp** [comment]
Just as an idea. 

Maybe it will help. In MockK I generate random value as matcher return value and repeat as many iterations as needed(because I have lambda as an argument to "when"). A number of iterations I take enough to distinguish values i.e. long=1, int=2, boolean=48. Then I gather all arguments for all iterations and create so-called "signatures" for arguments and compare them to a map of "signatures" for matchers. After matching that only constants are left and this is primitive values.

For Kotlin it is very important to have it working as many functions accept a lot of parameters with defined default values. Writing matcher per each argument is a big pain.

First I was focused on a similar approach you describing but then gave up few times till understood how to do it. I spent so much time thinking about how to do it, that after inventing it decided to write MockK framework. 

Of course this can have few disadvantages, but in reality really minor. Like what if argument value is time() function or random()

So may be for Mockito you can introduce special kind of "when()" construct with lambda, don't know.

**Comment by ChristianSchwarz** [comment]
@oleksiyp I came to a similar solution in [#828 "Prototype implementation of when syntax for void methods"](https://github.com/mockito/mockito/pull/828#issuecomment-400273820). Since `when` needs to accept a lambda the Mockito API needs an update, i guess this wont happen in Mockito 2. Introducing when with lambda would be useful to unify the whole stubbing API too.
